### DISCUSSION TITLE
DotNetCoreRestoreSettings doesn't support keys?

### DISCUSSION THREAD

**Discussion by mausch** [cake-build/cake#3434] [Q&A] [answered]
Hi! https://cakebuild.net/api/Cake.Common.Tools.DotNetCore/DotNetCoreAliases/6E238F91 doesn't mention any API keys.
I need to set API keys to restore from the GitHub package registry.

I browsed around and found https://cakebuild.net/api/Cake.Common.Tools.NuGet.Restore/NuGetRestoreSettings/ , that seems to be old/deprecated/duplicate?

Also found https://stackoverflow.com/a/38818942 which as a workaround manipulates nuget sources, but that's from 2016...

Many thanks

**Comment by augustoproiete** [comment, marked as answer]
`DotNetCoreRestore` is simply a wrapper around `dotnet restore` that comes with the .NET SDK.

`dotnet restore` does not support API keys directly. It relies on the existing NuGet sources that you set up previously on the machine (e.g. via `dotnet nuget`) and/or through a `nuget.config` file you can add to your repo with your nuget sources.

### Option 1: dotnet nuget add source

```powershell
dotnet nuget add source <nuget server> --name <name> --username xxx --password xxx --store-password-in-clear-text
```

If you wanted to call `dotnet nuget` via Cake, you can call via [`DotNetCoreNuGetAddSource`](
https://cakebuild.net/api/Cake.Common.Tools.DotNetCore/DotNetCoreAliases/450208ED) and can use [`DotNetCoreNuGetHasSource`](https://cakebuild.net/api/Cake.Common.Tools.DotNetCore/DotNetCoreAliases/A471A78D) to check if a source already exists before trying to create it. All the `dotnet nuget` Cake wrappers are documented here: https://cakebuild.net/dsl/dotnetcore/#NuGet

### Option 2: nuget.config

The other route is creating a [`nuget.config`](https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file) at the root of your repository that looks something like this:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="your-custom-feed" value="https://..." protocolVersion="3" />
  </packageSources>
  <activePackageSource>
    <add key="All" value="(Aggregate source)" />
  </activePackageSource>
  <packageSourceCredentials>
    <your-custom-feed>
      <add key="Username" value="xxx" />
      <add key="ClearTextPassword" value="xxx" />
    </your-custom-feed>
  </packageSourceCredentials>
</configuration>
```

---

- dotnet nuget add source documentation: https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file
- nuget.config documentation: https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file

**Comment by devlead** [comment]
The underlying SDK CLI dotnet restore command doesn't support specifying an API key
https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-restore#arguments

It can use either interactive or key present per source in implicit/specified nuget config file.

The dotnet nuget command can be used here i.e.

```bash
dotnet nuget add source --username USERNAME --password ${{ secrets.GITHUB_TOKEN }} --store-password-in-clear-text --name github "https://nuget.pkg.github.com/OWNER/index.json"
```

https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-nuget-registry#authenticating-to-github-packages


This is available in Cake using the DotNetCoreNuGetUpdateSource alias, example:

 ```cake
var settings = new DotNetCoreNuGetSourceSettings
 {
     Source = "https://www.example.com/nugetfeed",
     UserName = "username",
     Password = "password",
     StorePasswordInClearText = true
 };

 DotNetCoreNuGetUpdateSource("example", settings);
```

https://cakebuild.net/api/Cake.Common.Tools.DotNetCore/DotNetCoreAliases/F55178ED


If you're using github actions you can add token as a secret environment variable something like

```cake
steps:
  - name: Run Cake Script 
    run: dotnet cake 
    env:
      GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
```

And you can then access token as environment variable `GITHUB_TOKEN`

**Comment by mausch** [comment]
> dotnet restore does not support API keys directly. 

Ah yeah, every few months I trip on this and then forget that this is the underlying cause.
Thank you both :+1:

**Reply by nick5454** [reply]
I solved my issue by putting a Nuget,config in the build folder, so any nugets needed to run before i'm able to add sources also `
<configuration>
	<packageRestore>
		
		<add key="enabled" value="True" />

		
		<add key="automatic" value="True" />
	</packageRestore>

	
	<packageSources>
		<add key="NuGet official package source" value="https://api.nuget.org/v3/index.json" />
		<add key="smarts" value="https://pkgs.dev.azure.com/name/_packaging/name/nuget/v3/index.json" />
	</packageSources>

	
	<packageSourceCredentials>
		<smarts>
			<add key="Username" value="VSTS" />
			<add key="ClearTextPassword"  value="%NUGET_AUTH_TOKEN%" />
		</smarts>
	</packageSourceCredentials>

	
	<disabledPackageSources />

	
	

	

</configuration>`

THis works for me very well
